### ISSUE TITLE
error message: "Must specify file or stream output type"

### ISSUE THREAD

**Issue by arkyl** [PapenfussLab/gridss#160]
Hi,
I tested gridss on chr22 bam and its output seemed to claim that it finished the variant calling, but then it threw an error message about output.

`[Wed Aug 01 13:51:14 UTC 2018] gridss.CallVariants done. Elapsed time: 12.73 minutes.
Runtime.totalMemory()=5074583552
Exception in thread "main" java.lang.IllegalArgumentException: Must specify file or stream output type.
	at htsjdk.variant.variantcontext.writer.VariantContextWriterBuilder.build(VariantContextWriterBuilder.java:424)`

I specified the out using "O=". In the *.out.gridss.working dir, there is the *.out.breakpoint.vcf already generated. Any thoughts about this? Thanks a lot!

Yue

**Comment by d-cameron** [comment]
That errors comes from the htsjdk library not knowning what file format to write the output file in. Your output file must be in one of the supported file formats (.bcf or .vcf).

Adding a .vcf suffix to your output file will fix this error.

**Comment by arkyl** [comment]
Thanks very much. The .vcf suffix did fix the error. However the result *.out.vcf only contained the "SVTYPE=BND". There is no other SV type such as "DEL" or "INS" etc. 

I only run gridss on chr22 bam and it finished very fast. Thanks a lot for any help on this!

Yue

**Comment by d-cameron** [comment]
This is expected. GRIDSS uses[ VCF breakend notation](https://samtools.github.io/hts-specs/VCFv4.3.pdf#page=17) for all variants, including simple events such as deletions and insertions.

**Comment by arkyl** [comment]
Thanks very much for the explanation. I will try to install the StructuralVariantAnnotation and see how it goes. Thanks!
Yue

**Comment by arkyl** [comment]
Hi,
I tried to use the gridss.AnnotateVariants to annotate the breakend vcf, and I only got the SV calls that were annotated as BND (a total of 248 BND calls). It reduced from 19005 BND calls in the original output vcf but still no sv type such as DEL or INS. Is this expected? Thanks very much.

Yue

**Comment by arkyl** [comment]
Hi,
Sorry for all these questions. I tried to run the shell script gridss.sh and example_pipeline.sh in the example/; however, I somehow can't get right the ref genome hg19.fa. My version of hg19 is a mismatch to the chr12.1527326.DEL1024.bam. Can you provide the web_site/location that has the matching hg19? Thanks a lot!
Yue

**Comment by arkyl** [comment]
Hi,
I was testing gridss.sh:
I downloaded the hg19 and extracted only the chr sequences in the bam header of chr12.1527326.DEL1024.bam. 
The error of mismatch ref was gone; however, I still got the following error:

` ERROR   2018-08-14 19:22:29     CallVariants    Fatal exception thrown by worker thread.
java.lang.RuntimeException: Subprocess terminated with with exit status 139. Failed executing "bwa mem -t 8 /home/liuy/software_installed/gridss/example/chr12.1527326.DEL1024.bam.header.SN.fa ./chr12.1527326.DEL1024.bam.gridss.working/chr12.1527326.DEL1024.bam.realign.0.fq". Can you run the alignment command from the command line? Is the aligner on PATH? Did you build an index with prefix /home/liuy/software_installed/gridss/example/chr12.1527326.DEL1024.bam.header.SN.fa? `

The ref chr12.1527326.DEL1024.bam.header.SN.fa was bwa indexed. But the ./chr12.1527326.DEL1024.bam.gridss.working/chr12.1527326.DEL1024.bam.realign.0.fq was not in the dir chr12.1527326.DEL1024.bam.gridss.working/. Any ideas about what may cause this? Thanks a lot !

Yue

**Comment by d-cameron** [comment]
Two possible causes are: a) you didn't run picard tools ReorderSam on the bam^, or b) you didn't run `bwa index /home/liuy/software_installed/gridss/example/chr12.1527326.DEL1024.bam.header.SN.fa` before running GRIDSS.


^ GRIDSS requires the reference genome to match exactly. BAM files reference contig names by their ordinal and not their name. If it doesn't match exactly, ReorderSam can be used to rewrite the bam file to match the reference genome expected.

**Comment by arkyl** [comment]
Hi,
Thanks for the suggestions. The ref was bwa indexed. When I extracted the sequences from hg19, it was in the same order as in the bam header. 

To test it again, I used the ReorderSam to generate a new bam and it gave the same error:

`ERROR	2018-08-15 14:35:17	CallVariants	Fatal exception thrown by worker thread.
java.lang.RuntimeException: Subprocess terminated with with exit status 139. Failed executing "bwa mem -t 8 /home/liuy/software_installed/gridss/example/chr12.1527326.DEL1024.bam.header.SN.fa ./chr12.1527326.DEL1024.bam.reorder.bam.gridss.working/chr12.1527326.DEL1024.bam.reorder.bam.realign.0.fq". Can you run the alignment command from the command line? Is the aligner on PATH? Did you build an index with prefix /home/liuy/software_installed/gridss/example/chr12.1527326.DEL1024.bam.header.SN.fa?`

I think the problem is that there is no *.reorder.bam.realign.0.fq in the *.working/ dir. Thanks much!

Yue

**Comment by d-cameron** [comment]
If you add -Dgridss.keepTempFiles=true to your JVM command line arguments,
does the .realign.0.fq get generated? Do you have `bwa` in path? If so,
what version does it report?

If you're still having trouble, I recently released a GRIDSS docker image
(see
https://github.com/PapenfussLab/gridss/blob/master/docker/run_cohort_from_csv.sh
) which should solve most environment/configuration issues by isolating
GRIDSS and all dependencies from the host configuration.

**Comment by arkyl** [comment]
Hi,
Thanks for the reply. I can actually successfully run gridss-1.8.0-gridss-jar-with-dependencies.jar on a chr22 bam, and the whole genome bam generated from our sample mapped on hg38. I am still not sure why the example/ didn't work.  

I then tried to test the StructuralVariantAnnotation tool as it would give the actual SV types. I used simple-event-annotation.R from the gridss example/ as the template, on the small chr22 bam. It gave an error message but there was output which had the SIMPLE_TYPE(DEL/DUP/INS/INV/ITX) for the BND calls. The new vcf had the same number of entries as the input and the vast majority had SIMPLE_TYPE. I don't know if the the output is trustworthy. The error message is as the following:

`Warning message:
info fields with no header: SIMPLE_TYPE 
Warning message:
info fields with no header: SIMPLE_TYPE 
Error in data.frame(chrom = seqnames(simplegr), start = as.integer((start(simplegr) +  : 
  arguments imply differing number of rows: 0, 1
Execution halted`

I am also wondering if there is an alternative to generate the DEL/DUP/INS/INV/ITX calls from gridss output. If the algorithm of translating from BND to DEL/DUP/INS/INV/ITX is simple enough, I guess it might be easier and more flexible for me to write some script to do the translation to simply get the SIMPLE_TYPE.  Let me know what you think. Thanks very much for all your answers!

Yue
